Cesium CI Test

ggetz opened the pull request:
Test run for adding test to travis-ci.

- Installed Karma and additional karma plugins
- Configured karma
- Added task to gulp file for running karma
- Updated travis.yml to include running tests

mramato commented on Specs/test-main.js:
```diff
@@ -0,0 +1,55 @@
+    /* global require */
+    require({
+            // Karma serves files from '/base', which is the basePath from your config file
+            baseUrl : 'http://localhost:8080/base/Source',
```
This can just be `/base/Source`

mramato commented on Specs/test-main.js:
```diff
@@ -0,0 +1,55 @@
+    /*global __karma__*/
+    "use strict";
+    var allTestFiles = [];
+    var TEST_REGEXP = /(spec|test)\.js$/i;
```
I'm pretty sure you just want `/Spec\.js$/`, which covers the naming convention for all Cesium test files.  Anything else would allow for false positives.

mramato commented on Specs/test-main.js:
```diff
@@ -0,0 +1,55 @@
+    Object.keys(__karma__.files).forEach(function(file) {
+        if (TEST_REGEXP.test(file)) {
+            // Normalize paths to RequireJS module names.
+            allTestFiles.push(file.replace(/^\/base\//, 'http://localhost:8080/base/'));
```
This should just be `allTestFiles.push(file);`  Is there a reason you needed to hard code localhost and port?

In fact, I think the regex/forEach can be replaced with a simpler block:

```javascript
    var allTestFiles = Object.keys(__karma__.files).filter(function(file) {
        return /Spec\.js$/.test(file);
    });
```

ggetz replied:
This was a temporary fix to suppress [a warning](http://stackoverflow.com/questions/23706627/why-do-i-get-there-is-no-timestamp-for-base-src-tests-core-types-tests) I was getting for a lot of files.

mramato replied:
I'm not seeing those warnings, so maybe it's no longer needed.  Try it again after my suggested changes.

mramato commented on Specs/test-main.js:
```diff
@@ -0,0 +1,55 @@
+            paths : {
+                'Specs' : '../Specs',
+                'Source' : '.',
+                'Cesium' : 'Cesium'
```
This line shouldn't be needed.

mramato commented on karma.conf.js:
```diff
@@ -0,0 +1,96 @@
+    browsers: ['FirefoxWithWebGL'],
+
+
+    customLaunchers: {
```
I assume this was an effort to get WebGL working on Travis?  Or is this a workaround for running locally as well?

ggetz replied:
That was the attempt to get WebGL working in Travis, yes.

mramato commented:
While Travis is certainly a large part of the end goal, let's concentrate on getting 100% of the tests passing locally first.  Then we can worry about adjustments needed to support travis.

mramato commented on karma.conf.js:
```diff
@@ -0,0 +1,96 @@
+// Karma configuration
+// Generated on Mon Dec 14 2015 13:37:20 GMT-0500 (Eastern Standard Time)
```
You can get rid of these comments and also add the below to the top of the file to avoid jshint warnings:

Get rid

```javascript
/*jshint node:true*/
"use strict;"
```

mramato commented:
Let's rename `test-main.js` to `karma-main.js` to avoid confusion.  Let's also move `karma.conf.js` into the `Specs` directory.

mramato commented on karma.conf.js:
```diff
@@ -0,0 +1,88 @@
+
+    // list of files to exclude
+    exclude: [
+        'Specs/Widgets/**/*.js',
```
Let's remove these excludes so that all tests can be ran locally.  Later we can add category support to the runner so that we can exclude WebGL tests for Travis.

mramato replied:
Even if I remove these, I'm getting ` Executed 3649 of 7095`, do you know why it's not running them all?

ggetz replied:
Instead of just a failed test at the end, I get a 
````
Firefox 42.0.0 (Windows 8.1 0.0.0) ERROR
  TypeError: infoBox.frame.contentDocument is null
  at C:/workspace/cesium/Specs/Widgets/InfoBox/InfoBoxSpec.js:44
````
which is an error in the browser and not just a stack trace for a failed test. Maybe this error is crashing it?

mramato replied:
I don't think that's it, but that error is odd.  I just noticed that the customizeJasmine function is not being called until after some of the tests have run, but I'm not sure why. I think that's a big part of the problem.

ggetz replied:
If I run the test using Firefox, I get the same number as above, but if I run the tests using chrome, I get `Executed 4236 of 7095`

mramato commented on package.json:
```diff
@@ -44,9 +44,16 @@
     "gulp-rename": "1.2.2",
     "gulp-replace": "0.5.4",
     "gulp-zip": "3.0.2",
+    "jasmine-core": "^2.4.1",
```
Get rid of all of the carets in this file.  npm puts them there by default, but it's not good practice for production apps to use them.  We want to lock into specific versions for repeatability.

mramato commented on Specs/test-main.js:
```diff
@@ -0,0 +1,49 @@
+            deps : allTestFiles,
+
+            // we have to kickoff jasmine, as it is asynchronous
+            callback : __karma__.start
```
Okay, here's one of the big problems.  `callback` gets called before the below function.  We should get rid of this line and call `__karma__.start()` ourselves after customizeJasmine.

mramato commented on Specs/Core/requestAnimationFrameSpec.js:
```diff
@@ -29,8 +29,10 @@ defineSuite([
-                expect(callbackTimestamps[0]).toBeLessThanOrEqualTo(callbackTimestamps[1]);
-                expect(callbackTimestamps[1]).toBeLessThanOrEqualTo(callbackTimestamps[2]);
+                // We can't rely on the default matcher because this is called outside of jasmine
+                expect(callbackTimestamps[0] <= callbackTimestamps[1]).toBe(true);
```
So it turns out I was wrong and this is a symptom and not the cause of the problem.  I'm still trying to track down the real problem, but this code will ultimately be reverted.

mramato replied:
Let's revert this code ASAP because I think there are some fundamental problems that need to be addressed with how Jasmine is being run and patched.

mramato commented on gulpfile.js:
```diff
@@ -274,6 +275,19 @@ gulp.task('minifyRelease', ['generateStubs'], function() {
 
 gulp.task('release', ['combine', 'minifyRelease', 'generateDocumentation']);
 
+gulp.task('test', function(done) {
```
This should depend on `build`.

mramato commented on Specs/DataSources/GeoJsonDataSourceSpec.js:
```diff
@@ -697,7 +697,7 @@ defineSuite([
         return dataSource.load(mixedGeometries).then(function() {
             var entityCollection = dataSource.entities;
-            var entities = entityCollection.values;
+            var entities = entityCollection.entities;
```
I'm confused by this change.  This property doesn't exist.

mramato commented on karma.conf.js:
```diff
@@ -0,0 +1,90 @@
+/*jshint node:true*/
+"use strict;"
```
Type `"use strict;"` -> `"use strict";`  Also, add this to the list of files checked by jsHint (at the top of gulpfile.js).

mramato commented:
@ggetz can you give me write access to this branch so that I can submit the changes we talked about.  Thanks.

ggetz commented:
> @ggetz can you give me write access to this branch so that I can submit the changes we talked about. Thanks.

Ok, you should have push access.

mramato commented:
Thanks, my changes are in.

ggetz commented:
OK, thanks!

mramato commented on karma.conf.js:
```diff
@@ -0,0 +1,87 @@
+
+    // Concurrency level
+    // how many browser should be started simultanous
+    concurrency: Infinity
```
We can get rid of this field, since it's the default.  Let's also clean up the spacing in this file and run the auto-formatter on it.

ggetz replied:
How do you run the auto-formatter?

ggetz replied:
Never mind, I found it.

mramato commented on karma.conf.js:
```diff
@@ -0,0 +1,87 @@
+
+    // list of files to exclude
+    exclude: [
+        //'Specs/Widgets/**/*.js',
```
You can just make this an empty array, we won't actually exclude anything here.  It will get taken care of with our category system.

ggetz commented:
When I proxy the `/Data` directory to my `localhost:8080` and run `npm start,` I get all but one of the tests to run and only 1 fails.

mramato commented:
While that's probably not the final solution, it narrows down where potential problems might be.  Can you submit whatever you have so I can try it out.  I'm at the head of this branch but it crashes on the KML specs.

mramato commented on package.json:
```diff
@@ -70,6 +77,7 @@
     "minify": "gulp minify",
     "minifyRelease": "gulp minifyRelease",
     "release": "gulp release",
+    "test": "gulp test",
```
Here's what I envision as the standard test targets.  We might actually add more depending on developer needs, but I think our dev guide should explain how to tweak karma.conf.json for temporary debugging needs.

__test__ Run all tests with the default karma.conf.js browser (Chrome)
__test-all__ Run all tests with all browsers available on the system
__travis-test__ Run non-WebGL tests only with phantomjs and Firefox

We should also allow for a `--browsers` flag so the runner can specify the browsers in all cases.  I have some sample code for doing this.

mramato replied:
Actually, we can probably hold off on the `--browsers` flag if we are going to recommend devs just modifying the config file.

mramato commented on karma.conf.js:
```diff
@@ -0,0 +1,83 @@
+
+    // Continuous Integration mode
+    // if true, Karma captures browsers, runs the tests and exits
+    singleRun: false,
```
Let's set this to true by default, so you don't need to set it in the default `test` target.  This way the `test` target always matches whatever is in this file.

mramato commented on karma.conf.js:
```diff
@@ -0,0 +1,83 @@
+      suppressErrorSummary: true,
+      suppressFailed: false,
+      suppressPassed: true,  // do not print information about passed tests
+      suppressSkipped: true  // do not print information about skipped tests
```
Let's set these all to false by default and `maxLogLines` to 0.  This will provide more context for which tests are failing and what order they are being run in.  Since the tests are noisy, maxLogLines will cut down on all of the unnecessary output too.

mramato commented on .travis.yml:
```diff
@@ -2,5 +2,8 @@ language: node_js
+  - sh -e /etc/init.d/xvfb start
 script:
-  - npm run jsHint && npm run release
+  - npm run jsHint && npm run test && npm run release
```
We probably actually want to undo travis changes for the initial pull request (and not include a travis target either).  This will keep everything easier to manage.  1 pull request to get karma into Cesium and a second pull request to have karma run on travis.

ggetz commented:
> While that's probably not the final solution, it narrows down where potential problems might be. Can you submit whatever you have so I can try it out. I'm at the head of this branch but it crashes on the KML specs.

I submitted my changes. I run `npm start` and then `karma start karma.config.js --single-run` for it to work. For some reason it doesn't when I run it with `npm run test`

mramato commented:
> I submitted my changes. I run npm start and then karma start karma.config.js --single-run (which runs karma with Chrome) for it to work. For some reason it doesn't when I run it with npm run test (which runs it with Firefox)

This definitely helps narrow down the problem, but this isn't a good overall solution.  We don't want people to have to start a second server (the dev server) just to run the tests and we also don't want people to have to install karma globally or otherwise run it directly.  Now that you've pointed me in a good direction, I have some ideas.

ggetz commented:
> also don't want people to have to install karma globally or otherwise run it directly.

When I adjusted the gulp task to default to chrome, it also worked. So the problem is with Firefix, not with running karma globally.

mramato commented:
The core problem is a bug in karma: https://github.com/karma-runner/karma/issues/1070

The issue is that karma has a list of file extensions for determining binary files (which is a horrible approach to take).  When it serves up binary files with extensions it doesn't know about, they get corrupted.  I'm going to submit a workaround for the issue on our end, but that can be removed when the above bug is fixed. (We may even want to consider opening a PR to fix it ourselves).

mramato commented:
Okay, my fix is in.  The one test failure is actually fixed in upstream Cesium, so merge that in and all tests should pass!

ggetz commented:
Yes everything passes!

mramato commented:
:hand: High five!

ggetz commented:
:hand:

mramato commented:
As you might have noticed, there are now a ton of changes in this PR that we didn't make.  When you update a forked branch, it's a good idea to do it by going through your own master branch.

origin/cesium -> ggetz/master -> ggetz/cesium-ci.  This way this PR will reflect the PR as it will appear into the master Cesium repository.  You can easily fix this now by doing just that.

ggetz commented:
OK, I did that locally, I just didn't push. It should be fixed now.

mramato commented on .travis.yml:
```diff
@@ -2,5 +2,8 @@ language: node_js
 node_js:
   - "0.12"
 sudo: false
+before_script:
```
We can take this out for now too.

ggetz commented:
Thanks!

mramato commented:
There's only a few minor things to do before we can open the PR into Cesium (it can probably even happen today!)

* [ ] Add coverage support
* [ ] Add ie and safari launchers.
* [ ] Add `test-all` target for running on multiple browsers at once. (it might be worth checking out https://github.com/litixsoft/karma-detect-browsers, which I just found out about.)

mramato commented:
I'll end up duplicating this in the Cesium PR, but just for your own FYI.

After the initial karma support is in:
* [ ] Add back in category support. (this might be super simple)
* [ ] Port our old HTML reporter to the new system (if desired).
* [ ] Add `test-travis` target and make non-WebGL tests run as part of the CI process
* [ ] Delete `jscoverage` since it's no longer used.
* [ ] Update wiki for new test targets
* [ ] Add section about karma to testing guide.
* [ ] Remove test and coverage links from index.html and index.release.html
* [ ] Delete old Jasmine files and infrastructure

That very last link has other implications.  It will no longer be possible to run the tests via the website deployed version.  I personally don't feel this is a big deal, but I wanted to bring it up so that we were all on the same page and in case anyone else had concerns over this.

P.S.  The reviews may want some of the above things taken care of before we merge, but that's not a big deal either way.

mramato commented:
Okay, that's all I have.  Let me know when this is ready for another look.

pjcozzi commented:
> It will no longer be possible to run the tests via the website deployed version

Being able to run tests easily from previous releases sounds awesome, but we haven't really used it in practice so I think this is fine.

ggetz commented:
- Added ie and safari launchers
- Added detect-browser plugin
 - by default this is disabled in the `karma.config.js` file
- Added a `test-all` gulp target which enables the detect-browser plugin

Should I also add a phantomJS launcher in this PR?
